### PR TITLE
Update to latest Spin version with Factors

### PR DESCRIPTION
**Pull request by rylev**
This is ready for review, but it isn't quite ready to land as other PRs need to land first:

* [x] https://github.com/spinkube/containerd-shim-spin/pull/185
* [x] https://github.com/spinkube/spin-trigger-mqtt/pull/40
* [x] https://github.com/fermyon/spin-trigger-sqs/pull/48
* [x] https://github.com/fermyon/spin-trigger-command/pull/45

### CODE REVIEW

**Review by kate-goldenring — COMMENTED** [review]
This LGTM. This shows how the factor work greatly reduces the complexity of building Spin runtimes! I will be curious how e2e tests fair once this is marked Ready.

**Comment by Mossaka** [comment]
Need a rebase

File: containerd-shim-spin/src/engine.rs [hunk 1]
```diff
@@ -156,47 +159,53 @@ impl SpinEngine {
...
+                // this operation can be considered safe.
+                loader.enable_loading_aot_compiled_components();
+            },
+            // Currently, it is only possible to precompile applications distributed using
+            // `spin registry push`
+            Source::File(_) => {}
+        };
         let mut futures_list = Vec::with_capacity(trigger_types.len());
         for trigger_type in trigger_types.iter() {
+            let app = spin_app::App::new("TODO", app.clone());
```
[the remark this answers is not in the document]
Comment (rylev) [reply containerd-shim-spin/src/engine.rs:171]: This seems to be something we might be able to change in upstream as that app id is rarely actually used. Currently, in the Spin code base it's only being used in the Redis trigger for telemetry purposes. It's used by some other runtimes simply as a unique identifier.

File: conformance-tests/src/main.rs [hunk 2]
```diff
@@ -22,7 +22,11 @@ fn main() {
...
         .into();
-    conformance_tests::run_tests(move |test| run_test(test, &spin_binary, &ctr_binary)).unwrap();
+    let config = conformance_tests::Config::new("canary");
+    conformance_tests::run_tests(config, move |test| {
+        run_test(test, &spin_binary, &ctr_binary)
+    })
+    .unwrap();
 }
 
 fn run_test(
```
Comment (Mossaka) [inline conformance-tests/src/main.rs:32]: would be nice to add the conformance tests to the CI. Could be a separate PR though

File: containerd-shim-spin/src/engine.rs [hunk 3]
```diff
@@ -152,47 +155,52 @@ impl SpinEngine {
...
+                REDIS_TRIGGER_TYPE => {
                     info!(" >>> running spin redis trigger");
-                    redis_trigger.run(spin_trigger::cli::NoArgs)
+                    trigger::run::<RedisTrigger>(NoCliArgs, app, &loader).await?
                 }
-                SqsTrigger::TRIGGER_TYPE => {
-                    let sqs_trigger =
-                        build_trigger::<SqsTrigger>(app.clone(), app_source.clone()).await?;
+                SQS_TRIGGER_TYPE => {
                     info!(" >>> running spin sqs trigger");
```
Comment (Mossaka) [inline containerd-shim-spin/src/engine.rs:190]: suggest: These `info!()` could be moved to `run()` in `trigger.rs `

File: rust-toolchain.toml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 [toolchain]
-channel = "1.78.0"
+channel = "1.79.0"
```
Comment (Mossaka) [inline rust-toolchain.toml:2]: Could you please make this change to a separate PR?

File: containerd-shim-spin/src/engine.rs (hunk 3 above)
Comment (rylev) [reply containerd-shim-spin/src/engine.rs:190]: Hmmm... the `run` function is generic and handles all of the triggers. We could move the message into that function but it would likely look like ">>> running trigger \`spin_trigger_http::HttpTrigger\`" which I'm not sure is better.

File: rust-toolchain.toml (hunk 4 above)
Comment (rylev) [reply rust-toolchain.toml:2]: https://github.com/spinkube/containerd-shim-spin/pull/191

File: containerd-shim-spin/src/engine.rs (hunk 3 above)
Comment (Mossaka) [reply containerd-shim-spin/src/engine.rs:190]: Oh okay, I thought you can do `T::TYPE` to get the type of the trigger in `run()` function.

File: containerd-shim-spin/src/engine.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply containerd-shim-spin/src/engine.rs:171]: By globally do you mean unique to the cluster too? Container name may be a good option

**Comment by kate-goldenring** [comment]
@rylev can we mark this as "Ready for Review"?

File: containerd-shim-spin/src/engine.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rylev) [reply containerd-shim-spin/src/engine.rs:171]: The uniqueness isn't necessary (as far as I can tell) for everything to work properly - it is just needed for proper diagnostics as this string might be exposed to the user in telemetry. Therefore, I think we want something that makes sense to users if they see it. 

I'm not sure what the best thing to use is - if you say container name works well that's fine by me. How do I access that information?

**Comment by rylev** [comment]
@kate-goldenring I've marked this PR as ready to review, but we shouldn't merge until we've figured out the answer to [this thread](https://github.com/spinkube/containerd-shim-spin/pull/189#discussion_r1765479279).

File: containerd-shim-spin/src/engine.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply containerd-shim-spin/src/engine.rs:172]: We should be able to use the `HOSTNAME` env var to get the container name. This is not a unique container id; rather just the container name (ie `busybox`). Furthermore, if the container is using `hostNetwork`, the hostname becomes the node hostname, but in production, that is rarely done. For now `HOSTNAME` sounds like the best bet

[the remark this answers is not in the document]
Comment (kate-goldenring) [reply containerd-shim-spin/src/engine.rs:172]: Did a quick check locally, printing the value of `HOSTNAME` from the shim and looks like it is the fully unique container name: `"container name: simple-spinapp-59c4d9668-gz8dr"`

[the remark this answers is not in the document]
Comment (kate-goldenring) [reply containerd-shim-spin/src/engine.rs:172]: @rylev what are your thoughts on using `HOSTNAME`?

[the remark this answers is not in the document]
Comment (rylev) [reply containerd-shim-spin/src/engine.rs]: I don't have a better idea. I suppose this isn't hard to change in the future and since nothing is really relying on the id being a particular format (it's just to help humans), we don't really even need to worry about backwards compatibility.

File: containerd-shim-spin/src/engine.rs [hunk 5]
```diff
@@ -150,49 +153,48 @@ impl SpinEngine {
...
+            // Currently, it is only possible to precompile applications distributed using
+            // `spin registry push`
+            Source::File(_) => {}
+        };
+
         let mut futures_list = Vec::new();
         let mut trigger_type_map = Vec::new();
-
+        // The `HOSTNAME` environment variable should contain the fully unique container name
+        let app_id = std::env::var("HOSTNAME").unwrap_or_else(|| "unknown".into());
```
Comment (kate-goldenring) [inline containerd-shim-spin/src/engine.rs:172]: Cargo clippy has some thoughts in CI
```suggestion
        let app_id = std::env::var("HOSTNAME").unwrap_or_else(|_| "unknown".into());
```

**Comment by kate-goldenring** [comment]
@rylev I am not sure why the MQTT test is failing. I've tested it locally and it passes. I am rerunning the tests to see if it is a race case with how the test is written

**Comment by kate-goldenring** [comment]
@rylev I there was an issue with spin.toml parsing from the latest changes to the MQTT trigger  that is causing mqtt tests to fail: https://github.com/spinkube/spin-trigger-mqtt/pull/40.

This should be the fix https://github.com/spinkube/spin-trigger-mqtt/pull/41

**Review by kate-goldenring — APPROVED** [review]
🎉

**Review by Mossaka — APPROVED** [review]
